### ISSUE TITLE
Make `-ea` builds of the latest development release available

### ISSUE THREAD

**Issue by sxa** [adoptium/temurin-build#3355]
At the moment we create nightly builds of the latest codebase for the current in-development release (JDK21 at the time of writing). https://jdk.java.net/21/  provides builds on the equvivalent of our primary platforms for each new build level (tag) that is in the openjdk repository, along with the list of changes and issues associated with that build, so it is a "known" level that people can use to test preview features without the additional risk of picking up a nightly.

We should look at making the ea levels available to customers, perhaps initially by either utilising the weekly runs, or alternatively by running an ad-hoc build from the tag whenever it is detected. This can then be communicated to the users interested in picking up those builds on all of the platforms we target. As per normal process, we would still be marking these as "beta" or equivalent as they would be not undergo TCK testing, and in fact we would not generally make any claims as to the quality of these builds, although I would anticipate running it through the test suites so they are visible to anyone that wanted them.

While the above would be the minimum I'd want to get in place, some potential future items could be:
- Make this accessible separately through the API (Unforunately `ea` in the API currently means nightly, but maybe we could have something else?)
- Make them available on the web site
- Do increased levels of triage on the test results
- Anything else?

**Comment by sxa** [comment]
Noting that if you run a pipeline with `overridePublishName` at the moment and try to release it it falls foul of the regex checks in the release tool e.g. 

`Processing non-archive file: OpenJDK-debugimage_aarch64_linux_hotspot_21_20-sxaea.tar.gz`
`ERROR: OpenJDK-debugimage_aarch64_linux_hotspot_21_20-sxaea.tar.gz is an archive but does not match regex OpenJDK([[:digit:]]*)U?(-jre|-jdk|-debugimage|-static-libs-glibc|-static-libs|-static-libs-musl|-testimage|-jdk-sources)_([[:alnum:]\-]+_)?([[:alnum:]\-]+_)?([[:alnum:]\-_]+_)?([[:digit:]\-]+|[[:alnum:]\._]+)\.(tar\.gz|zip|pkg|msi)$, please update sbin/Release.sh`

(That file name was produced with a publishname of `jdk-21+20-sxaea` - we'd also need to decide if we want the output from the build to be different from a normal nightly to indicate its status as a "fixed" ea level based on a tag.

**Comment by sxa** [comment]
Without modifying the release tool it is not possible to achieve this since there is a regex check for non-RELEASE builds that insists on the timestamp being present in the filename before it will publish to github.

We can release a "nightly without publish" build under a release that has the `TAG` set to e.g. `jdk-21+21-ea` and something suitable in the timestamp field for the filenames e.g. `2023-05-25-00-00` which will give a release like this: https://github.com/adoptium/temurin21-binaries/releases/tag/jdk-21%2B21-ea-beta

**Comment by sxa** [comment]
JDK21+22:
- https://ci.adoptium.net/job/build-scripts/job/openjdk21-pipeline/86/
- https://ci.adoptium.net/job/build-scripts/job/release/job/refactor_openjdk_release_tool/6523/
- Visible at https://github.com/adoptium/temurin21-binaries/releases/tag/jdk-21%2B22-ea-beta
JDK21+23 (plus `--with-version-pre=ea-beta` and standard nightly testing
- https://ci.adoptium.net/job/build-scripts/job/openjdk21-pipeline/87/
-

**Comment by sxa** [comment]
So TL;DR (Details/alternatives can be read in previous comments if desired):
Proposal to build and publish the EA levels:
- Run openjdk21-pipeline with:
  -  `configureArgs` set to `--with-version-opt=ea` to give a `-ea` on the end of the build number (See previous comment for what this does to the output)
  - The one hotspot configuration removed from `targetConfigurations` so it only builds VARIANT=temurin
  - Set `scmReference` to the appropriate adopt tag e.g. `jdk-21+23_adopt`
  - `releaseType` = `Nightly Without Publish`
- [Run the release tool](https://ci.adoptium.net/job/build-scripts/job/release/job/refactor_openjdk_release_tool/build) with:
  -  `tag`: `jdk-21+23-ea`
  - A timestamp such as `2023-05-25-00-99` (Not ideal, but as mentioned before `99` will make it look different for the ea builds - we'd need to modify https://github.com/adoptium/github-release-scripts/blob/ba2f0167d794c1cf29314292265045453a8ce7bc/sbin/Release.sh#L32 to allow an alternate format for something that isn't a "release" build)

**Comment by sxa** [comment]
Noting that the `ea`  builds, being prefixed jdk-21 instead of jdk-2023, are showing up at the top of the list before the nightlies.

**Comment by sxa** [comment]
Anyone wishing to use the beta builds with e.g. `setup-java` in github actions can use a URL of e.g. https://api.adoptium.net/v3/binary/version/jdk-21+22-ea-beta/linux/aarch64/jre/hotspot/normal/eclipse

For example:
```
    steps:
      - name: Checkout
        uses: actions/checkout@v3

      - name: pwd
        run: |
          pwd

      - name: Download latest JDK21
        run: |
          curl -L -o jdk.tar.gz https://api.adoptium.net/v3/binary/version/jdk-21+22-ea-beta/linux/x64/jdk/hotspot/normal/eclipse

      - uses: actions/setup-java@v3
        with:
          jdkFile: jdk.tar.gz
          java-version: '21'
          java-package: jdk
          architecture: x64
          distribution: 'jdkfile'
```

**Comment by sxa** [comment]
Schedule on the openjdk21-pipeline is now removed and a schedule has been put on the https://ci.adoptium.net/job/build-scripts/job/utils/job/releaseTrigger_21ea/ job to run once a day if there is a new tag that hasn't yet been published. The publish step will still need to be run manually (Fixable later by a pipeline script that can determine the correct `UPSTREAM_BUILD_NUMBER` to pass to the release tool

**Comment by sxa** [comment]
jdk-21+29:

- Build trigger: https://ci.adoptium.net/job/build-scripts/job/utils/job/releaseTrigger_21ea/20/console
- Main platforms: [build openjdk21-pipeline#134](https://ci.adoptium.net/job/build-scripts/job/openjdk21-pipeline/134) | [Release job](https://ci.adoptium.net/job/build-scripts/job/release/job/refactor_openjdk_release_tool/6713/)
- Evaluation platforms: [evaluation-openjdk21-pipeline](https://ci.adoptium.net/job/build-scripts/job/evaluation-openjdk21-pipeline/21) | [Release job](https://ci.adoptium.net/job/build-scripts/job/release/job/refactor_openjdk_release_tool/6714/)
- Published release: https://github.com/adoptium/temurin21-binaries/releases/tag/jdk-21%2B29-ea-beta

**Comment by sxa** [comment]
jdk-21+30:
- Main platforms: [openjdk21-pipeline](https://ci.adoptium.net/job/build-scripts/job/openjdk21-pipeline/144/) | [Release job](https://ci.adoptium.net/job/build-scripts/job/release/job/refactor_openjdk_release_tool/6769/)
- Evaluation platforms: No successful one available
- Published release

**Comment by sxa** [comment]
jdk-21+31: https://ci.adoptium.net/job/build-scripts/job/openjdk21-pipeline/153/parameters/ | https://ci.adoptium.net/job/build-scripts/job/release/job/refactor_openjdk_release_tool/6825/ | [Published release](https://github.com/adoptium/temurin21-binaries/releases/tag/jdk-21%2B30-ea-beta)

**Comment by sxa** [comment]
Non-evaluation builds are now self-publishing to https://github.com/adoptium/temurin21-binaries/releases?q=jdk-21&expanded=true

**Comment by sxa** [comment]
Closing this issue and will create a follow-up issue for further comments/suggestions.

**Comment by sxa** [comment]
Noting that for the ea builds other than the original the automation was trying to push with ea_21.0.1-0-12
Removing the extra zero and converting `.` to `-` characters  so that it has a `TIMESTAMP` parameter of `ea_21-0-1-12` works as desired.
